File: apps/sdk/examples/python/kubercoin_example.py
#!/usr/bin/env python3
"""
KuberCoin Python SDK Example
Demonstrates how to interact with the KuberCoin RPC API
"""

from typing import Any, Dict, List, Optional
import requests


class KuberCoinRpcError(RuntimeError):
    """Raised when the RPC endpoint returns an application-level error."""


class KuberCoinClient:
    """Simple KuberCoin RPC client"""

    def __init__(self, rpc_url: str = "http://localhost:8634/"):
        self.rpc_url = rpc_url
        self.request_id = 0
        self.timeout_seconds = 10

    def _call(self, method: str, params: Optional[List[Any]] = None) -> Any:
        """Make an RPC call"""
        self.request_id += 1

        payload = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params or [],
            "id": self.request_id,
        }

        response = requests.post(
            self.rpc_url,
            json=payload,
            headers={"Content-Type": "application/json"},
            timeout=self.timeout_seconds,
        )
        response.raise_for_status()

        result = response.json()

        if "error" in result and result["error"]:
            raise KuberCoinRpcError(f"RPC Error: {result['error']}")

        return result.get("result")

    def get_block_count(self) -> int:
        """Get current blockchain height"""
        return self._call("getblockcount")

    def get_best_block_hash(self) -> str:
        """Get hash of the best (tip) block"""
        return self._call("getbestblockhash")

    def get_block(self, block_hash: str) -> Dict:
        """Get block by hash"""
        return self._call("getblock", [block_hash])

    def get_block_by_height(self, height: int) -> Dict:
        """Get block by height (resolves hash first, then fetches block)"""
        block_hash = self._call("getblockhash", [height])
        return self._call("getblock", [block_hash])

    def get_transaction(self, txid: str) -> Dict:
        """Get transaction by txid (verbose)"""
        return self._call("getrawtransaction", [txid, True])

    def get_mempool_info(self) -> Dict:
        """Get mempool information"""
        return self._call("getmempoolinfo")

    def get_peer_info(self) -> List[Dict]:
        """Get connected peer information"""
        return self._call("getpeerinfo")


def main():
    """Example usage"""
    print("KuberCoin Python SDK Example")
    print("=" * 50)

    # Initialize client
    client = KuberCoinClient()

    try:
        # Get blockchain info
        height = client.get_block_count()
        print(f"\n📊 Blockchain Height: {height}")

        best_hash = client.get_best_block_hash()
        print(f"🔗 Best Block Hash: {best_hash}")

        # Get latest block
        print("\n📦 Latest Block Details:")
        latest_block = client.get_block_by_height(height)
        latest_height = latest_block.get("height")
        latest_hash = latest_block.get("hash")
        latest_prev = latest_block.get("previousblockhash")
        latest_txs = latest_block.get("tx") or []
        latest_nonce = latest_block.get("nonce")
        print(f"   Height: {latest_height}")
        print(f"   Hash: {latest_hash}")
        print(f"   Previous: {latest_prev}")
        print(f"   Transactions: {len(latest_txs)}")
        print(f"   Nonce: {latest_nonce}")

        # Get genesis block
        print("\n🌱 Genesis Block:")
        genesis = client.get_block_by_height(0)
        genesis_hash = genesis.get("hash")
        genesis_time = genesis.get("time")
        print(f"   Hash: {genesis_hash}")
        print(f"   Time: {genesis_time}")

        # Get mempool info
        print("\n💾 Mempool Info:")
        mempool = client.get_mempool_info()
        print(f"   Size: {mempool.get('size')} transactions")
        print(f"   Bytes: {mempool.get('bytes')}")

        # Get first transaction from latest block
        if latest_block["tx"]:
            first_tx = latest_block["tx"][0]
            print("\n💸 First Transaction in Block:")
            tx = client.get_transaction(first_tx)
            print(f"   TxID: {tx.get('txid')}")
            print(f"   Inputs: {len(tx.get('vin') or [])}")
            print(f"   Outputs: {len(tx.get('vout') or [])}")

        print("\n✅ All operations completed successfully!")

    except (KuberCoinRpcError, requests.RequestException, ValueError, KeyError) as e:
        print(f"\n❌ Error: {e}")


if __name__ == "__main__":
    main()

File: apps/sdk/examples/javascript/kubercoin_example.js
/**
 * KuberCoin JavaScript/Node.js SDK Example
 * Demonstrates how to interact with the KuberCoin RPC API
 */

import { pathToFileURL } from 'node:url';

class KuberCoinClient {
    constructor(rpcUrl = 'http://localhost:8634/') {
        this.rpcUrl = rpcUrl;
        this.requestId = 0;
    }

    /**
     * Make an RPC call
     */
    async call(method, params = []) {
        this.requestId++;
        
        const payload = {
            jsonrpc: '2.0',
            method: method,
            params: params,
            id: this.requestId
        };

        const response = await fetch(this.rpcUrl, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const result = await response.json();

        if (result.error) {
            throw new Error(`RPC Error: ${JSON.stringify(result.error)}`);
        }

        return result.result;
    }

    /**
     * Get current blockchain height
     */
    async getBlockCount() {
        return await this.call('getblockcount');
    }

    /**
     * Get hash of the best (tip) block
     */
    async getBestBlockHash() {
        return await this.call('getbestblockhash');
    }

    /**
     * Get block by hash
     */
    async getBlock(blockHash) {
        return await this.call('getblock', [blockHash]);
    }

    /**
     * Get block by height (resolves hash first, then fetches block)
     */
    async getBlockByHeight(height) {
        const hash = await this.call('getblockhash', [height]);
        return await this.call('getblock', [hash]);
    }

    /**
     * Get transaction by txid (verbose)
     */
    async getTransaction(txid) {
        return await this.call('getrawtransaction', [txid, true]);
    }

    /**
     * Get mempool information
     */
    async getMempoolInfo() {
        return await this.call('getmempoolinfo');
    }

    /**
     * Get connected peer information
     */
    async getPeerInfo() {
        return await this.call('getpeerinfo');
    }
}

/**
 * Example usage
 */
// Run if called directly (Node.js)
const isMain = process.argv[1] && pathToFileURL(process.argv[1]).href === import.meta.url;
if (isMain) {
    console.log('KuberCoin JavaScript SDK Example');
    console.log('='.repeat(50));

    const client = new KuberCoinClient();

    try {
        // Get blockchain info
        const height = await client.getBlockCount();
        console.log(`\n📊 Blockchain Height: ${height}`);

        const bestHash = await client.getBestBlockHash();
        console.log(`🔗 Best Block Hash: ${bestHash}`);

        // Get latest block
        console.log(`\n📦 Latest Block Details:`);
        const latestBlock = await client.getBlockByHeight(height);
        console.log(`   Height: ${latestBlock.height}`);
        console.log(`   Hash: ${latestBlock.hash}`);
        console.log(`   Previous: ${latestBlock.previousblockhash}`);
        console.log(`   Transactions: ${latestBlock.tx.length}`);
        console.log(`   Nonce: ${latestBlock.nonce}`);

        // Get genesis block
        console.log(`\n🌱 Genesis Block:`);
        const genesis = await client.getBlockByHeight(0);
        console.log(`   Hash: ${genesis.hash}`);
        console.log(`   Time: ${new Date(genesis.time * 1000).toISOString()}`);

        // Get mempool info
        console.log(`\n💾 Mempool Info:`);
        const mempool = await client.getMempoolInfo();
        console.log(`   Size: ${mempool.size} transactions`);
        console.log(`   Bytes: ${mempool.bytes}`);

        // Get first transaction from latest block
        if (latestBlock.tx.length > 0) {
            const firstTx = latestBlock.tx[0];
            console.log(`\n💸 First Transaction in Block:`);
            const tx = await client.getTransaction(firstTx);
            console.log(`   TxID: ${tx.txid}`);
            console.log(`   Inputs: ${tx.vin.length}`);
            console.log(`   Outputs: ${tx.vout.length}`);
        }

        console.log('\n✅ All operations completed successfully!');
    } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        console.error(`\n❌ Error: ${message}`);
    }
}

// Export for use in other modules
export { KuberCoinClient };
